Ignore Backport Properties in SBOL Comparison

GeneCodeSavvy opened the pull request:
While writing tests for **Attachment** objects, I encountered a bug related to namespace properties during SBOL document conversions:

* Converting an **SBOL2** document to **SBOL3** adds a new `namespace` property.
* When converting the resulting **SBOL3** document back to **SBOL2**, a **backport namespace** property is added.
* This causes test failures because `sbol_diff.file_diff()` performs a strict one-to-one file comparison, and the added backport property results in mismatches.

This PR ensures that backport properties are ignored during file comparisons, preventing false test failures.

Related to #357

GeneCodeSavvy committed:
add files for sbol_diff tests

GeneCodeSavvy committed:
add new function to remove backport triplets from the rdf graph

GeneCodeSavvy committed:
add tests for file_diff ignore backport properties

jakebeal reviewed (changes requested):
I very much like this and think that it is a very useful contribution. 

I would like to suggest adding one additional tweak, however: let the function take an argument to say whether it is stripping out SBOL2 or SBOL3 backport properties, and then strip out only those.

The reason for this is to make sure that backport properties are properly cleaned up after a round trip. If we have an SBOL2 file, then we want to strip SBOL3 backport properties, but should flag a problem if there is an SBOL2 backport property still in the file. Likewise, an SBOL3 file may have SBOL2 backport properties, but should not have any SBOL3 backport properties.

Gonza10V commented:
Agree with Jake, seems reasonable and is not difficult to implement. After this change we should be able to detect if the objects have unintended backport properties.

GeneCodeSavvy committed:
add validation of backport properties

GeneCodeSavvy committed:
make detect_sbol_version function more robust

GeneCodeSavvy committed:
add test for sbol detect_sbol_version

GeneCodeSavvy committed:
fix backport property logic for .nt files

GeneCodeSavvy committed:
add more tests for sbol version detection

GeneCodeSavvy committed:
add test for diff method

GeneCodeSavvy committed:
test files for diff tests

GeneCodeSavvy commented:
I have added functions to detect the sbol version from the files, and error out if the the document contains invalid backport properties. Therefore,  Catching mistakes in the convertor

bbartley commented:
@GeneCodeSavvy  @Gonza10V  Correct me if I'm wrong, but does this solution permanently alter the files being diffed by stripping them of their backport annotations?  Wouldn't we want to restore those triples after performing the diff?

Gonza10V commented:
@bbartley Hmm, I think we dont use those files afterwards, the strip should be just for tests. If an user would like to strip backport properties and then use the file maybe should make a copy of the file and process one and then pass the other.

bbartley commented:
@Gonza10V I'm concerned because `sbol_diff.file_diff()` is part of a general utility package, and it's entirely possible that other users might use it for different purposes than our round-trip tests.  Also, a `diff` operation that modifies its input files is very unusual, and could lead to some nasty surprises for unwitting users.  So, I would suggest _not_ making this the default behavior.  Rather, we could add an optional flag such as `strip_backport_annotations=False` that preserves the default behavior, but then have the round-trip tests use `strip_backport_annotations=True`

Gonza10V commented:
I like the proposal. The diff function should not change the files that is comparing. The function should have `strip_backport_annotations` as parameter, and be False by default so users are aware of the stripping. Let me know your thoughts.

GeneCodeSavvy commented:
No function created in this PR affect the original files. The Graph object created from the files are being manipulated, in memory and are not saved.
cc @Gonza10V @bbartley

bbartley commented:
@GeneCodeSavvy sorry, you are right that _the files are not changed_ -- rather I should have said _the Graphs are changed_ .  But the same point applies -- a diff operation that modifies the comparands is unusual and unexpected, so we need some safeguard so that users are not caught unaware.  Also, there may be use cases where a user wants to perform a diff and then continue to use the Graphs unmodified afterwards.

GeneCodeSavvy committed:
fix bug - function modifies graph object

GeneCodeSavvy committed:
support flag for strip_backport_properties while diffing sbol files

GeneCodeSavvy committed:
update test files for sbol_diff

GeneCodeSavvy commented:
Thank you pointing that out @bbartley. I have changed the function, so that graphs are also not changed. I have also added a new flag `strip_backport_properties` that is false by default, so it should not break previous uses and users have to consciously set the flag to true to ignore the backport properties while diffing.

Gonza10V commented:
@jakebeal  can you take a look to this? this looks good to me and you can merge if it looks good to you.

jakebeal commented on sbol_utilities/sbol_diff.py:
```diff
@@ -16,9 +18,120 @@ def _load_rdf(fpath: Union[str, bytes, os.PathLike]) -> rdflib.Graph:
+    for _, p, _ in graph:
+        # Check predicates for SBOL namespace URIs
+        predicate_str = str(p)
+        if sbol2_namespace in predicate_str:
```
These should be testing the prefix with `startswith` rather than `in`.
It's unlikely to go wrong given the constraints on URIs, but it's good code hygiene.

blaze27444 replied:
Outdated

jakebeal commented on sbol_utilities/sbol_diff.py:
```diff
@@ -16,9 +18,120 @@ def _load_rdf(fpath: Union[str, bytes, os.PathLike]) -> rdflib.Graph:
+    sbol3_namespace = 'http://sbols.org/v3#'
+
+    # Check bound namespaces first
+    for _, namespace in graph.namespaces():
```
If a document contains both sbol2 and sbol3 namespaces, this iteration will produce unpredictable results, depending on the order of iteration through the graph. Instead of an iteration, use a set comprehension to collect all the namespaces. Then if both are present, return `None`, and if only one is present return that one.

The same change should be made to search of predicates below.

blaze27444 replied:
Outdated

jakebeal commented on sbol_utilities/sbol_diff.py:
```diff
@@ -16,9 +18,120 @@ def _load_rdf(fpath: Union[str, bytes, os.PathLike]) -> rdflib.Graph:
+    return clean_graph
+
+
+def _diff_graphs(g1: rdflib.Graph, g2: rdflib.Graph,
```
Function needs a docstring

blaze27444 replied:
Outdated

jakebeal commented on sbol_utilities/sbol_diff.py:
```diff
@@ -16,9 +18,120 @@ def _load_rdf(fpath: Union[str, bytes, os.PathLike]) -> rdflib.Graph:
+    g2_sbol_version = _detect_sbol_version(g2)
+
+    if g1_sbol_version != g2_sbol_version:
+        raise TypeError('SBOL documents are of different versions, cannot compare')
```
I see no reason that we can't compare documents of different versions - they just won't be equal!

blaze27444 replied:
Outdated

jakebeal commented on sbol_utilities/sbol_diff.py:
```diff
@@ -16,9 +18,120 @@ def _load_rdf(fpath: Union[str, bytes, os.PathLike]) -> rdflib.Graph:
+        raise TypeError('SBOL documents are of different versions, cannot compare')
+
+    if g1_sbol_version:
+        validate_backport_properties(g1, g1_sbol_version)
```
Why are we validating in a comparison function? This feels inappropriate.
In test functions, we should be comparing separately from validating backport properties.

blaze27444 replied:
Outdated

jakebeal commented on sbol_utilities/sbol_diff.py:
```diff
@@ -16,9 +18,120 @@ def _load_rdf(fpath: Union[str, bytes, os.PathLike]) -> rdflib.Graph:
+    if g2_sbol_version:
+        validate_backport_properties(g2, g2_sbol_version)
+
+    g1_to_compare = g1
```
I think that everything above this point is not needed in this function.

blaze27444 replied:
Outdated

jakebeal commented on test/test_sbol_diff.py:
```diff
@@ -51,6 +63,129 @@ def test_doc_diff(self):
+        version3 = sbol_utilities.sbol_diff._detect_sbol_version(graph3_xml)
+        self.assertEqual('sbol3', version3, 'Should detect SBOL3 document')
+
+        sbol3_file_nt = os.path.join(TEST_FILES_DIR, 'sbol3_collection.nt')
```
Add a test for a document with mixed SBOL2 and SBOL3 namespaces. You can make this by hand-editing a copy of `sbol3_collection.nt` to be an invalid mixed file.

jakebeal commented on test/test_sbol_diff.py:
```diff
@@ -51,6 +63,129 @@ def test_doc_diff(self):
+        clean_graph = sbol_utilities.sbol_diff._load_rdf(clean_file)
+
+        result = sbol_utilities.sbol_diff.validate_backport_properties(clean_graph, 'sbol2')
+        self.assertTrue(result, 'Clean SBOL2 document should pass validation')
```
Also test an invalid document.

blaze27444 replied:
Outdated

jakebeal commented on test/test_sbol_diff.py:
```diff
@@ -51,6 +63,129 @@ def test_doc_diff(self):
+        version3_nt = sbol_utilities.sbol_diff._detect_sbol_version(graph3_nt)
+        self.assertEqual('sbol3', version3_nt, 'Should detect SBOL3 document')
+
+    def test_cross_version_comparison_raises_error(self):
```
Per my note in `sbol_diff.py`, I think this should just be "false" rather than "error"

blaze27444 replied:
Outdated

jakebeal reviewed (changes requested):
This is very nice work - just a few minor tweaks needed to finish, I think.

GeneCodeSavvy committed:
refactor detect_sbol_version function

GeneCodeSavvy committed:
add doc strings to _diff_graph function

GeneCodeSavvy committed:
drop unnecessary code in the _diff_graphs function

GeneCodeSavvy committed:
update doc-strings

GeneCodeSavvy committed:
fix small bug with strict comparision

GeneCodeSavvy committed:
fix bug : namespace checking is not robust enough

GeneCodeSavvy committed:
update tests

GeneCodeSavvy commented:
I have made the changes as per your comments @jakebeal

Commit:
Merge branch 'develop' into feature/ignore_backport_properties_sbol_diff

jakebeal commented on sbol_utilities/sbol_diff.py:
```diff
@@ -18,35 +19,32 @@ def _load_rdf(fpath: Union[str, bytes, os.PathLike]) -> rdflib.Graph:
 
 
-def _detect_sbol_version(graph: rdflib.Graph) -> Optional[str]:
+def _detect_sbol_version(graph: rdflib.Graph):
```
```suggestion
def _detect_sbol_version(graph: rdflib.Graph) -> Optional[str]:
```
Why delete the type information?

blaze27444 replied:
Outdated

jakebeal commented on sbol_utilities/sbol_diff.py:
```diff
@@ -18,35 +19,32 @@ def _load_rdf(fpath: Union[str, bytes, os.PathLike]) -> rdflib.Graph:
+    if has_v2 and has_v3:
+        return None
+    elif has_v2:
+        return 'sbol2'
```
Consider changing these "magic value" constants (both here and elsewhere in the code) into values from a string enum type with a name like SBOLVersion. Using values like "SBOLVersion.SBOL2" instead of raw string values will allow type-checking to be applied to the values and ensure that they are used consistently across different functions.

blaze27444 replied:
Outdated

jakebeal commented on sbol_utilities/sbol_diff.py:
```diff
@@ -112,18 +108,25 @@ def _remove_selective_backport_properties(graph: rdflib.Graph) -> rdflib.Graph:
+             - in_g1: Statements that are only in the first graph.
+             - in_g2: Statements that are only in the second graph.
+    """
     g1_sbol_version = _detect_sbol_version(g1)
```
These version values aren't used any more and should be deleted.

blaze27444 replied:
Outdated

jakebeal reviewed (approved):
Nice work!

GeneCodeSavvy committed:
replace constant strings with enums

GeneCodeSavvy committed:
remove usused version

GeneCodeSavvy commented:
@jakebeal  please review
